http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala

```diff
@@ -800,12 +800,12 @@ class LeonardoService(protected val dataprocConfig: DataprocConfig,
     } yield {
       // Making the assumption that users will always be able to access clusters that they create
       // Fix for https://github.com/DataBiosphere/leonardo/issues/821
-      val visibleClusters = samVisibleClusters ::: clusterList
```

Comment:
Here's the bug. We were essentially `:::`ing a `List[(GoogleProject, ClusterInternalId)]` with a `List[(GoogleProject, ClusterName)]`. This compiled, and the inferred type was `List[(GoogleProject, Product with Serializable)]`. We were then doing a `.contains` check on that, which was failing.

I reorganized the code a bit and tried to make the types explicit. The compiler now complains if we pass the wrong type to `.contains`.

Reply:
not enough type safety :)

Reply:
Yup